### PR TITLE
DataResourceType now throws instead of returning Result

### PR DESCRIPTION
**Pull request by lucianomarisi**

### CODE REVIEW

File: Sources/TABResourceLoader/Protocols/Resource types/DataResourceType.swift [hunk 1]
```diff
@@ -16,5 +16,5 @@ public protocol DataResourceType: ResourceType {
 
    - returns: An instantiated model if parsing was succesful, otherwise nil
    */
-  func result(from data: Data) -> Result<Model>
+  func result(from data: Data) throws -> Model
```
Comment (samdods) [inline Sources/TABResourceLoader/Protocols/Resource types/DataResourceType.swift]: the name `result(from:)` is technically no longer correct. i think it should be `model(from:)`

File: Sources/TABResourceLoader/Protocols/Resource types/JSONArrayResourceType.swift [hunk 2]
```diff
@@ -24,19 +24,18 @@ public protocol JSONArrayResourceType: DataResourceType {
...
+      throw JSONParsingError.notAJSONArray
     }
     do {
-      let parsedResult = try model(from: jsonArray)
-      return .success(parsedResult)
+      return try model(from: jsonArray)
     } catch {
-      return .failure(error)
+      throw error
     }
```
Comment (samdods) [inline Sources/TABResourceLoader/Protocols/Resource types/JSONArrayResourceType.swift]: i think this `do/catch` is pointless. the containing method throws, so you can just do `return model(from: jsonArray)`

File: Sources/TABResourceLoader/Protocols/Resource types/JSONDictionaryResourceType.swift [hunk 3]
```diff
@@ -24,20 +24,19 @@ public protocol JSONDictionaryResourceType: DataResourceType {
...
+      throw JSONParsingError.notAJSONDictionary
     }
 
     do {
-      let parsedResult = try model(from: jsonDictionary)
-      return .success(parsedResult)
+      return try model(from: jsonDictionary)
     } catch {
-      return .failure(error)
+      throw error
```
Comment (samdods) [inline Sources/TABResourceLoader/Protocols/Resource types/JSONDictionaryResourceType.swift]: same comment as above

**Review by samdods — APPROVED** [review]
approved pending comments being addressed

File: Sources/TABResourceLoader/Protocols/Resource types/DataResourceType.swift (hunk 1 above)
Comment (lucianomarisi) [reply Sources/TABResourceLoader/Protocols/Resource types/DataResourceType.swift]: makes sense, I'll rename

File: Sources/TABResourceLoader/Protocols/Resource types/JSONArrayResourceType.swift (hunk 2 above)
Comment (lucianomarisi) [reply Sources/TABResourceLoader/Protocols/Resource types/JSONArrayResourceType.swift]: you're correct I'll remove

File: Sources/TABResourceLoader/Services/NetworkResponseHandler.swift [hunk 4]
```diff
@@ -28,6 +28,11 @@ struct NetworkResponseHandler {
       return .failure(NetworkServiceError.noData)
     }
 
-    return resource.result(from: data)
+    do {
+      let model = try resource.model(from: data)
+      return .success(model)
+    } catch {
+      return .failure(NetworkServiceError.networkingError(error: error))
```
Comment (deanbrindley87) [inline Sources/TABResourceLoader/Services/NetworkResponseHandler.swift]: is this right to be throwing network service error here even though the network request was probably ok, how come we're not just returning `.failure(error)`?

**Review by deanbrindley87 — APPROVED** [review]

File: Sources/TABResourceLoader/Services/NetworkResponseHandler.swift (hunk 4 above)
Comment (lucianomarisi) [reply Sources/TABResourceLoader/Services/NetworkResponseHandler.swift]: good point, I'll change to `.failure(error)` here

**Pull request merged**